libbeat/publisher/queue/queue.go

```diff
@@ -80,23 +80,34 @@ type ProducerConfig struct {
 	DropOnCancel bool
 }
 
-// Producer interface to be used by the pipelines client to forward events to be
-// published to the queue.
-// When a producer calls `Cancel`, it's up to the queue to send or remove
-// events not yet ACKed.
-// Note: A queue is still allowed to send the ACK signal after Cancel. The
-//       pipeline client must filter out ACKs after cancel.
+// Producer is an interface to be used by the pipelines client to forward
+// events to a queue.
 type Producer interface {
+	// Publish adds an event to the queue, blocking if necessary, and returns
+	// true on success.
 	Publish(event publisher.Event) bool
+
+	// TryPublish adds an event to the queue, returning immediately if the queue
+	// is full, and returns true on success.
 	TryPublish(event publisher.Event) bool
+
+	// Cancel closes this Producer endpoint. If the producer is configured to
+	// drop its events on Cancel, the number of dropped events is returned.
+	// Note: A queue may still send ACK signals even after Cancel is called on
+	//       the originating Producer. The pipeline client must accept and
+	//       discard these ACKs.
```

Comment:
If I recall correctly the client based cancellers are closed as well. communication is async, which means that 'during' cancel we still get ACKs, but once close has finished no ACKs might be received anymore. The pipeline also has a 'global' ACK queue, which will still receive the ACKs.

Reply:
Could you clarify? I'm not sure what "client based cancellers" are in this context. This comment was based on the code in `{memqueue, spool}/produce.go:Close()` which only closes the `done` channel of the producer's `openState`. Unless you're talking about the `producerCancel{Request, Response}` handling when `dropOnCancel` is set?

Reply:
I see. Yeah, you are right here. The behavior about ACKs is in the beat.Client interface, not the Producer interface.